File: experiments/haiku-subagents/friction.py
#!/usr/bin/env python3
"""Classify per-arm request friction into reads / accepted-writes / rejected-writes /
auth-redirects — so polling is no longer conflated with rejection.

Two log shapes:
  proxy  — /tmp/arena-<arm>.http.jsonl  ({method,path,status} per HTTP request)
  erpc   — /tmp/erpc-game.jsonl         (event_type per RPC; needs state_read logging)

    uv run --no-project experiments/haiku-subagents/friction.py proxy /tmp/arena-proto.http.jsonl
    uv run --no-project experiments/haiku-subagents/friction.py erpc  /tmp/erpc-game.jsonl [gameId]
"""
import sys
import json

KIND = sys.argv[1]
PATH = sys.argv[2]
GAME = sys.argv[3] if len(sys.argv) > 3 else None

reads = writes_ok = writes_rej = auth = other = 0

with open(PATH) as f:
    for line in f:
        line = line.strip()
        if not line:
            continue
        try:
            e = json.loads(line)
        except ValueError:
            continue
        if KIND == "proxy":
            method, status = e.get("method"), e.get("status", 0)
            if status == 302:
                auth += 1                        # bounced to /login (session lost / no cookie)
            elif method == "GET":
                reads += 1                        # poll / read (200, or 304)
            elif method == "POST" and status == 303:
                writes_ok += 1                    # accepted move, Proto PRG
            elif method == "POST" and status == 200:
                writes_ok += 1                    # accepted move, Simple
            elif method == "POST" and status >= 400:
                writes_rej += 1                   # rejected move (400 malformed / 403 rule)
            else:
                other += 1
        elif KIND == "erpc":
            if GAME and e.get("game_id") != GAME:
                continue
            et = e.get("event_type")
            if et == "state_read":
                reads += 1
            elif et == "move_accepted":
                writes_ok += 1
            elif et == "move_rejected":
                writes_rej += 1
            # player_assigned / game_over are neither read nor write-attempt

writes = writes_ok + writes_rej
ratio = f"{reads / writes:.1f}:1" if writes else "n/a"
print(f"{PATH.split('/')[-1]}")
print(f"  reads (poll):     {reads}")
print(f"  writes accepted:  {writes_ok}")
print(f"  writes rejected:  {writes_rej}")
print(f"  auth redirects:   {auth}")
if other:
    print(f"  other:            {other}")
print(f"  read:write ratio: {ratio}   (rejections are NOT counted as reads)")

File: experiments/oss-driver/Friction.fs
module TicTacToe.OssDriver.Friction

// Classify per-arm request friction into reads / accepted-writes / rejected-writes /
// auth-redirects, so polling is not conflated with rejection. Ported from friction.py.
//   proxy log: one {ts,method,path,status} per line (the F# proxy subcommand's output)
//   erpc  log: one {event_type,game_id,...} per line
//
//   dotnet run --project experiments/oss-driver -- friction proxy /tmp/arena-surface.http.jsonl
//   dotnet run --project experiments/oss-driver -- friction erpc  /tmp/erpc-game.jsonl [gameId]

open System.IO
open System.Text.Json

let run (argv: string[]) : int =
    if argv.Length < 2 then
        eprintfn "usage: friction <proxy|erpc> <log> [gameId]"
        2
    else
        let kind = argv.[0]
        let path = argv.[1]
        let game = if argv.Length > 2 then Some argv.[2] else None
        let mutable reads, ok, rej, auth, other = 0, 0, 0, 0, 0
        for raw in File.ReadLines path do
            let line = raw.Trim()
            if line <> "" then
                try
                    use doc = JsonDocument.Parse line
                    let root = doc.RootElement
                    let getS (name: string) =
                        match root.TryGetProperty name with
                        | true, v -> v.GetString()
                        | _ -> null
                    let getI (name: string) =
                        match root.TryGetProperty name with
                        | true, v -> (try v.GetInt32() with _ -> 0)
                        | _ -> 0
                    if kind = "proxy" then
                        let m, status = getS "method", getI "status"
                        if status = 302 then auth <- auth + 1
                        elif m = "GET" then reads <- reads + 1
                        elif m = "POST" && (status = 303 || status = 200) then ok <- ok + 1
                        elif m = "POST" && status >= 400 then rej <- rej + 1
                        else other <- other + 1
                    elif kind = "erpc" then
                        let skip = game.IsSome && getS "game_id" <> game.Value
                        if not skip then
                            match getS "event_type" with
                            | "state_read" -> reads <- reads + 1
                            | "move_accepted" -> ok <- ok + 1
                            | "move_rejected" -> rej <- rej + 1
                            | _ -> ()
                with _ -> ()
        let writes = ok + rej
        let ratio = if writes > 0 then sprintf "%.1f:1" (float reads / float writes) else "n/a"
        printfn "%s" (Path.GetFileName path)
        printfn "  reads (poll):     %d" reads
        printfn "  writes accepted:  %d" ok
        printfn "  writes rejected:  %d" rej
        printfn "  auth redirects:   %d" auth
        if other > 0 then printfn "  other:            %d" other
        printfn "  read:write ratio: %s   (rejections are NOT counted as reads)" ratio
        0
